packages/devtools_app/lib/src/shared/console/eval/eval_service.dart

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:vm_service/vm_service.dart';
+
+import '../../../service/vm_service_wrapper.dart';
+import '../../object_tree.dart';
+import '../primitives/eval_history.dart';
+
+class AutocompleteCache {
+  final classes = <ClassRef, Class>{};
+
+  /// Cache of autocomplete matches for a library for code written within that
+  /// library.
+  ///
+  /// This cache includes autocompletes from all libraries imported and exported
+  /// by the library as well as all private autocompletes for the library.
+  final libraryMemberAndImportsAutocomplete =
+      <LibraryRef, Future<Set<String?>>>{};
+
+  /// Cache of autocomplete matches to show for a library when that library is
+  /// imported.
+  ///
+  /// This cache includes autocompletes from libraries exported by the library
+  /// but does not include autocompletes for libraries imported by this library.
+  final libraryMemberAutocomplete = <LibraryRef, Future<Set<String?>>>{};
+
+  void clear() {
+    classes.clear();
+    libraryMemberAndImportsAutocomplete.clear();
+    libraryMemberAutocomplete.clear();
+  }
+}
+
+class EvalService {
+  EvalService({
+    required this.isolateRef,
+    required this.variables,
+    required this.frameForEval,
+    required this.isPaused,
+    required this.service,
+  });
+
+  final IsolateRef? Function() isolateRef;
+
+  final bool Function() isPaused;
+
+  final Frame? Function() frameForEval;
+
+  final List<DartObjectNode> Function() variables;
+
+  final VmServiceWrapper Function() service;
+
+  final EvalHistory evalHistory = EvalHistory();
+
+  final cache = AutocompleteCache();
+
+  String get _isolateRefId {
+    final id = isolateRef()?.id;
+    if (id == null) return '';
```

Comment:
why is returning '' for the isolateRefId ever ok?

Reply:
I do not know. I copied it from previous code. Added 'TODO'.